File: app/src/main/java/com/shumidub/todoapprealm/realmcontrollers/notescontroller/FolderNotesRealmController.java
package com.shumidub.todoapprealm.realmcontrollers.notescontroller;


import com.shumidub.todoapprealm.App;
import com.shumidub.todoapprealm.realmcontrollers.RealmDb;
import com.shumidub.todoapprealm.realmmodel.notes.FolderNotesObject;
import com.shumidub.todoapprealm.realmmodel.notes.NoteObject;

import io.realm.RealmList;

public class FolderNotesRealmController {

    //FolderNotes

    public static FolderNotesObject getFolderNote(long id) {
        return RealmDb.findById(FolderNotesObject.class, id);
    }

    public static long addFolderNote(String name){
        long id = RealmDb.newUniqueId(FolderNotesObject.class);
        RealmDb.write(() -> {
            FolderNotesObject folder = RealmDb.realm().createObject(FolderNotesObject.class);
            folder.setId(id);
            folder.setName(name);
            App.folderOfNotesContainerList.add(folder);
        });
        return id;
    }

    public static void editFolderNote(long id, String name){
        RealmDb.write(() -> getFolderNote(id).setName(name));
    }

    public static void delFolderNote(long id){
        RealmList<NoteObject> realmList = getFolderNote(id).getTasks();
        RealmDb.write(() -> {
            realmList.deleteAllFromRealm();
            RealmDb.realm().where(FolderNotesObject.class).equalTo("id", id).findAll().deleteAllFromRealm();
        });
    }


    public static void reorderFolderNote(int from, int to){
        RealmDb.write(() ->
                App.folderOfNotesContainerList.add(to, App.folderOfNotesContainerList.remove(from)));
    }

    public static long getNewValidFolderNotesId() {
        return RealmDb.newUniqueId(FolderNotesObject.class);
    }

    //Notes

    public static RealmList<NoteObject> getNotesList(long idFolderNotesObject){
        return getFolderNote(idFolderNotesObject).getTasks();
    }

    public static NoteObject getNote(long idNotesObject){
        return RealmDb.findById(NoteObject.class, idNotesObject);
    }

    public static long addNote(long idFolderNotesObject, String text){
        long id = RealmDb.newUniqueId(NoteObject.class);
        FolderNotesObject folderNotesObject = getFolderNote(idFolderNotesObject);
        RealmDb.write(() -> {
            NoteObject noteObject = RealmDb.realm().createObject(NoteObject.class);
            noteObject.setId(id);
            noteObject.setText(text);
            noteObject.setIdFolder(idFolderNotesObject);
            folderNotesObject.getTasks().add(noteObject);
        });
        return id;
    }

    public static void editNote(long idNotesObject, String text ){
        RealmDb.write(() -> getNote(idNotesObject).setText(text));
    }

    public static void delNote(long idNotesObject){
        RealmDb.write(() -> {
            NoteObject noteObject = getNote(idNotesObject);
            FolderNotesObject folderNotesObject = getFolderNote(noteObject.getIdFolder());
            folderNotesObject.getTasks().remove(noteObject);
            noteObject.deleteFromRealm();
        });
    }

    public static void reorderNote(long idFolderNotesObject, long idNotesObject, int from, int to){
        RealmList<NoteObject> notesList = getFolderNote(idFolderNotesObject).getTasks();
        RealmDb.write(() -> notesList.add(to, notesList.remove(from)));
    }

    public static long getNewValidNotesId() {
        return RealmDb.newUniqueId(NoteObject.class);
    }
}

File: app/src/main/java/com/shumidub/todoapprealm/realmcontrollers/notescontroller/FolderNotesRealmController.kt
package com.shumidub.todoapprealm.realmcontrollers.notescontroller

import com.shumidub.todoapprealm.App
import com.shumidub.todoapprealm.realmmodel.RealmFoldersContainer
import com.shumidub.todoapprealm.realmmodel.notes.FolderNotesObject
import com.shumidub.todoapprealm.realmmodel.notes.NoteObject
import io.realm.kotlin.ext.query

object FolderNotesRealmController {

    // Folders

    fun getFoldersList(): List<FolderNotesObject> =
        App.realm.query<RealmFoldersContainer>().first().find()
            ?.folderOfNotesList
            ?.toList()
            .orEmpty()

    fun getFolderNote(id: Long): FolderNotesObject? =
        App.realm.query<FolderNotesObject>("id == $0", id).first().find()

    fun addFolderNote(name: String): Long {
        val newId = getNewValidFolderNotesId()
        App.realm.writeBlocking {
            val container = query<RealmFoldersContainer>().first().find() ?: return@writeBlocking
            val folder = copyToRealm(FolderNotesObject().apply {
                this.id = newId
                this.name = name
            })
            container.folderOfNotesList.add(folder)
        }
        return newId
    }

    fun editFolderNote(id: Long, name: String) {
        App.realm.writeBlocking {
            val folder = query<FolderNotesObject>("id == $0", id).first().find()
                ?: return@writeBlocking
            folder.name = name
        }
    }

    fun delFolderNote(id: Long) {
        App.realm.writeBlocking {
            val folder = query<FolderNotesObject>("id == $0", id).first().find()
                ?: return@writeBlocking
            folder.notesObjectRealmList.toList().forEach { delete(it) }
            delete(folder)
        }
    }

    fun reorderFolderNote(from: Int, to: Int) {
        App.realm.writeBlocking {
            val list = query<RealmFoldersContainer>().first().find()?.folderOfNotesList
                ?: return@writeBlocking
            if (from !in list.indices || to !in list.indices) return@writeBlocking
            val item = list.removeAt(from)
            list.add(to, item)
        }
    }

    // Notes

    fun getNotesList(folderId: Long): List<NoteObject> =
        App.realm.query<FolderNotesObject>("id == $0", folderId).first().find()
            ?.notesObjectRealmList
            ?.toList()
            .orEmpty()

    fun getNote(id: Long): NoteObject? =
        App.realm.query<NoteObject>("id == $0", id).first().find()

    fun addNote(folderId: Long, text: String): Long {
        val newId = getNewValidNotesId()
        App.realm.writeBlocking {
            val folder = query<FolderNotesObject>("id == $0", folderId).first().find()
                ?: return@writeBlocking
            val note = copyToRealm(NoteObject().apply {
                this.id = newId
                this.text = text
                this.idFolder = folderId
            })
            folder.notesObjectRealmList.add(note)
        }
        return newId
    }

    fun editNote(id: Long, text: String) {
        App.realm.writeBlocking {
            val note = query<NoteObject>("id == $0", id).first().find() ?: return@writeBlocking
            note.text = text
        }
    }

    fun delNote(id: Long) {
        App.realm.writeBlocking {
            val note = query<NoteObject>("id == $0", id).first().find() ?: return@writeBlocking
            delete(note)
        }
    }

    fun reorderNote(folderId: Long, noteId: Long, from: Int, to: Int) {
        App.realm.writeBlocking {
            val list = query<FolderNotesObject>("id == $0", folderId).first().find()
                ?.notesObjectRealmList ?: return@writeBlocking
            if (from !in list.indices || to !in list.indices) return@writeBlocking
            val item = list.removeAt(from)
            list.add(to, item)
        }
    }

    private fun getNewValidFolderNotesId(): Long {
        var id = System.currentTimeMillis()
        while (App.realm.query<FolderNotesObject>("id == $0", id).first().find() != null) {
            id++
        }
        return id
    }

    private fun getNewValidNotesId(): Long {
        var id = System.currentTimeMillis()
        while (App.realm.query<NoteObject>("id == $0", id).first().find() != null) {
            id++
        }
        return id
    }
}
